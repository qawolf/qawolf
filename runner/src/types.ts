import { Frame, Page } from "playwright";

export type Artifacts = {
  gifUrl: string | null;
  logsUrl: string;
  videoUrl: string | null;
};

export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id?: string;
  version: number;
};

export type Run = {
  artifacts?: Artifacts;
  code: string;
  env?: Record<string, string>;
  helpers: string;
  id: string;
  test_id: string;
  version: number;
};

export type RunOptions = {
  artifacts?: Artifacts;
  code: string;
  end_line?: number;
  env?: Record<string, string>;
  helpers: string;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id: string;
  version: number;
};

export type RunHook = {
  after?(result: RunProgress): Promise<void>;
  progress?(progress: RunProgress): Promise<void>;
  before?(): Promise<void>;
};

export type RunStatus = "created" | "fail" | "pass";

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number;
  error?: string;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
  test_id: string;
};

export type Suite = {
  env?: Record<string, string>;
  runs: Run[];
};

export type VideoChapter = {
  lineCode: string;
  lineNum: number;
  start: number;
};

export type BrowserName = "chromium" | "firefox" | "webkit";

export type Callback<S = void, T = void> = (data?: S) => T;

export type ElementAction = "click" | "fill" | "press" | "selectOption";

export interface ElementEvent {
  action: ElementAction;
  page: Page;
  frame?: Frame;
  frameSelector?: string;
  selector: string;
  value?: string | null;
}

export type WindowAction = "goBack" | "goto" | "popup" | "reload";

export interface WindowEvent {
  action: WindowAction;
  page: Page;
  popup?: Page;
  value?: string | null;
}

export interface LogEvent {
  level: string;
  message: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Variables = { [key: string]: any };
