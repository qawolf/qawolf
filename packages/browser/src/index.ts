// public API
export { BrowserContext, register } from "./ContextManager";
export { Page } from "./page/PageManager";
// TODO move to jest-playwright
export { launch, LaunchResult } from "./launch";

// internal API
export * from "./actions";
export * from "./keyboard";
export { ContextManager } from "./ContextManager";
export { PageManager } from "./page/PageManager";
