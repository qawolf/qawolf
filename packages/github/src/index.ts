import Octokit from "@octokit/rest";

const octokit = new Octokit({
  auth: "d9e1b19607b2b07ed5c6f2e78fa5fb1bd3cabf26"
});

const main = async () => {
  const res = await octokit.checks.create({
    owner: "qawolf",
    repo: "internal",
    name: "QA Wolf",
    head_sha: "9c52a7a6f2090c58bcb23ae066ce31920789f71c"
  });

  console.log("RESPONSE", res);
};

main();
