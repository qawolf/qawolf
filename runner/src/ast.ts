import ts from "typescript";

export const buildStatements = (
  sourceCode: string
): ts.NodeArray<ts.Statement> =>
  ts.createSourceFile("source.ts", sourceCode, ts.ScriptTarget.Latest)
    .statements;
