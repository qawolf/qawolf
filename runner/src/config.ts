import { v4 as uuidv4 } from "uuid";

export const DEFAULT_ATTRIBUTE_LIST =
  "data-cy,data-e2e,data-qa,/^data-test.*/,/^qa-.*/";

export const buildRunnerApiKey = (): string => {
  const uuid = uuidv4().split("-").join("");
  return `qawolf_${uuid}`;
};

const RUNNER_ID = process.env.QAWOLF_RUNNER_ID;

const config = {
  API_URL: process.env.QAWOLF_API_URL || "https://app.qawolf.com/api",
  DEFAULT_ATTRIBUTE_LIST,
  // 1080 viewport + 100 browser
  DISPLAY_HEIGHT: process.env.QAWOLF_DISPLAY_HEIGHT || 1180,
  // 1280 viewport + 8 scrollbar
  DISPLAY_WIDTH: process.env.QAWOLF_DISPLAY_WIDTH || 1288,
  FFMPEG_PATH: process.env.FFMPEG_PATH || "/usr/bin/ffmpeg",
  INTERNAL_SERVER_PORT: 26368,
  RECORDER_SCRIPT_FILENAME: "qawolf.recorder.js",
  RUNNER_API_KEY: RUNNER_ID ? buildRunnerApiKey() : undefined,
  RUNNER_ID,
  RUNNER_STATUS_URL:
    process.env.QAWOLF_RUNNER_STATUS_URL ||
    "http://localhost/.qawolf/runner/status",
};

export default config;
