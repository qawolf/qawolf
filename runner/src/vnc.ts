import { execSync } from "child_process";
import { promises as fs } from "fs";

const VNC_DIR = "/root/.vnc";
const PASSWD_PATH = `${VNC_DIR}/passwd`;

export const setVncPassword = async (password: string): Promise<void> => {
  try {
    await fs.mkdir(VNC_DIR);
  } catch (e) {
    if (e.code !== "EEXIST") {
      throw e;
    }
  }

  execSync(`vncpasswd -f <<< ${password} > "${PASSWD_PATH}"`, {
    shell: "/bin/bash",
  });

  await fs.chmod(PASSWD_PATH, "600");
};
