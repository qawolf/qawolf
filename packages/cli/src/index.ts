#!/usr/bin/env node

import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { yellow } from "kleur";
import program from "commander";
import { snakeCase } from "lodash";
import { saveCiTemplate } from "./ci";
import { record } from "./record";
import { test } from "./test";
import { parseUrl } from "./utils";
const { version } = require("../package");

program.version(version);

program.usage("<command> [options]");

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "save events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  logger.verbose(`record url "${url.href}"`);

  const name = snakeCase(optionalName || url.hostname!);
  await record(url, name, cmd.events);
});

program
  .command("test [name]")
  .description("run a test")
  .action(async optionalName => {
    await test(optionalName);
  });

program
  .command("azure")
  .description("set up an Azure Pipeline")
  .action(async () => {
    await saveCiTemplate("azure");
  });

program
  .command("circleci")
  .description("set up CircleCI")
  .action(async () => {
    await saveCiTemplate("circleci");
  });

program
  .command("github")
  .description("set up a GitHub Action")
  .action(async () => {
    await saveCiTemplate("github");
  });

program
  .command("gitlab")
  .description("set up GitLab CI/CD")
  .action(async () => {
    await saveCiTemplate("gitlab");
  });

program.arguments("<command>").action(cmd => {
  console.log(yellow(`Invalid command "${cmd}"\n`));
  program.help();
});

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
