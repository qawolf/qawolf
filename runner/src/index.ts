import Debug from "debug";
import fs from "fs";
import path from "path";

import config from "./config";
import { launch } from "./environment/launch";
import { RunnerServer } from "./server/RunnerServer";
import { StatusReporter } from "./services/StatusReporter";
import { setVncPassword } from "./vnc";

const debug = Debug("qawolf:index");

const { access } = fs.promises;

const recorderScriptPath = require.resolve(
  path.join(__filename, `../../${config.RECORDER_SCRIPT_FILENAME}`)
);

export const warmUp = async (): Promise<void> => {
  debug("warm up");
  process.env.DISPLAY = ":0.0";
  // force the browser cold start (0.5s to 15s)
  // https://github.com/microsoft/playwright/issues/4345
  const { browser, context } = await launch({ headless: false });
  // create and navigate a page to prevent an issue where
  // the first page opened has an issue resolving a url
  const page = await context.newPage();
  await page.goto("https://google.com");
  await browser.close();
};

export const verifyRecorderScriptExists = async (): Promise<void> => {
  try {
    await access(recorderScriptPath);
  } catch (error) {
    debug(
      `QA Wolf recorder script not found at "${path.resolve(
        recorderScriptPath
      )}". `,
      "Did you do `npm run build-recorder-script`?"
    );
    process.exit(1);
  }
};

export const start = async (): Promise<RunnerServer> => {
  await verifyRecorderScriptExists();

  await setVncPassword(config.RUNNER_API_KEY || "local");

  // warm up the session before starting the server or
  // notifying that the session is ready
  await warmUp();

  const server = await RunnerServer.start();
  const reporter = new StatusReporter(server._runner);

  process.on("SIGTERM", () => {
    reporter.close();
    server.close();
  });

  return server;
};

if (require.main === module) {
  start();
}
