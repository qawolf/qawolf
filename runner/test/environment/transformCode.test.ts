import { buildStatements } from "../../src/ast";
import {
  getExpressions,
  getVariableNames,
  transformCode,
} from "../../src/environment/transformCode";

const simpleSourceCode = 'await page.click("a");';

const sourceCode = `

await page.click(
  '[placeholder="Search by location.."]'
);
await page.fill('[placeholder="Search by location.."]', "Lakeside Village");
await page.click("a");

// comment about function
function myFunction() {
  const myInnerConst = "hi";
  console.log("SUP"!);
}

myFunction();

await myFunction();

var myVar = "hello";
const myConst =  "world";
let myLet = "!";
myLet = "!!";
`;

const sourceCode2 = `const todo = "hi there!";
await page.goto('http://todomvc.com/examples/react');

// enter new todo
await page.click(".new-todo");

await page.fill(".new-todo", todo);
await page.press(".new-todo", "Enter");

// complete todo
await page.click(
    ".toggle"
);

// clear completed
await page.click(".clear-completed");`;

describe("getExpressions", () => {
  it("parses expressions from code", () => {
    expect(getExpressions(simpleSourceCode)).toMatchSnapshot();
    expect(getExpressions(sourceCode)).toMatchSnapshot();
    expect(getExpressions(sourceCode2)).toMatchSnapshot();
  });
});

describe("getVariableName", () => {
  const getCodeVariableNames = (code: string, statement = 0): string[] => {
    const statements = buildStatements(code);
    return getVariableNames(statements[statement]);
  };

  it("parses top level variables", () => {
    expect(getCodeVariableNames(`var a = "1";`)).toEqual(["a"]);
  });

  it("parses nested variables", () => {
    expect(getCodeVariableNames(`var { b, c } = { b: 2, c: 3 };`)).toEqual([
      "b",
      "c",
    ]);
  });

  it("parses multiple declarations", () => {
    expect(
      getCodeVariableNames(`var [d, e], [f, g] = [4, 5], [6, 7];`)
    ).toEqual(["d", "e", "f", "g"]);
  });

  it("parses redeclarations", () => {
    expect(getCodeVariableNames(`var a = 0;\na = "2";`, 1)).toEqual(["a"]);
  });
});

describe("transformCode", () => {
  it("transforms web editor code", () => {
    expect(
      transformCode({
        code: simpleSourceCode,
        variables: { page: "page" },
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode,
        variables: { page: "page", x: 1, y: null, z: "hi" },
      })
    ).toMatchSnapshot();
  });

  it("takes start and end lines into account", () => {
    expect(
      transformCode({
        code: simpleSourceCode,
        endLine: 1,
        startLine: 1,
        variables: { page: "page" },
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode,
        endLine: 5,
        startLine: 4,
        variables: { page: "page", x: 1, y: null, z: "hi" },
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode,
        endLine: 12,
        startLine: 2,
        variables: { page: "page", x: 1, y: null, z: "hi" },
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode2,
        endLine: 4,
        startLine: 1,
        variables: {},
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode2,
        endLine: 8,
        startLine: 6,
        variables: {},
      })
    ).toMatchSnapshot();

    expect(
      transformCode({
        code: sourceCode2,
        endLine: 15,
        startLine: 10,
        variables: {},
      })
    ).toMatchSnapshot();
  });
});
