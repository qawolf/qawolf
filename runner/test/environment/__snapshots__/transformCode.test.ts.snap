// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExpressions parses expressions from code 1`] = `
Array [
  Object {
    "code": "await page.click(\\"a\\");",
    "endLine": 1,
    "startLine": 1,
    "variableNames": Array [],
  },
]
`;

exports[`getExpressions parses expressions from code 2`] = `
Array [
  Object {
    "code": "await page.click(
  '[placeholder=\\"Search by location..\\"]'
);",
    "endLine": 5,
    "startLine": 3,
    "variableNames": Array [],
  },
  Object {
    "code": "await page.fill('[placeholder=\\"Search by location..\\"]', \\"Lakeside Village\\");",
    "endLine": 6,
    "startLine": 6,
    "variableNames": Array [],
  },
  Object {
    "code": "await page.click(\\"a\\");",
    "endLine": 7,
    "startLine": 7,
    "variableNames": Array [],
  },
  Object {
    "code": "// comment about function
function myFunction() {
  const myInnerConst = \\"hi\\";
  console.log(\\"SUP\\"!);
}",
    "endLine": 13,
    "startLine": 9,
    "variableNames": Array [],
  },
  Object {
    "code": "myFunction();",
    "endLine": 15,
    "startLine": 15,
    "variableNames": Array [],
  },
  Object {
    "code": "await myFunction();",
    "endLine": 17,
    "startLine": 17,
    "variableNames": Array [],
  },
  Object {
    "code": "var myVar = \\"hello\\";",
    "endLine": 19,
    "startLine": 19,
    "variableNames": Array [
      "myVar",
    ],
  },
  Object {
    "code": "const myConst =  \\"world\\";",
    "endLine": 20,
    "startLine": 20,
    "variableNames": Array [
      "myConst",
    ],
  },
  Object {
    "code": "let myLet = \\"!\\";",
    "endLine": 21,
    "startLine": 21,
    "variableNames": Array [
      "myLet",
    ],
  },
  Object {
    "code": "myLet = \\"!!\\";",
    "endLine": 22,
    "startLine": 22,
    "variableNames": Array [
      "myLet",
    ],
  },
]
`;

exports[`getExpressions parses expressions from code 3`] = `
Array [
  Object {
    "code": "const todo = \\"hi there!\\";",
    "endLine": 1,
    "startLine": 1,
    "variableNames": Array [
      "todo",
    ],
  },
  Object {
    "code": "await page.goto('http://todomvc.com/examples/react');",
    "endLine": 2,
    "startLine": 2,
    "variableNames": Array [],
  },
  Object {
    "code": "// enter new todo
await page.click(\\".new-todo\\");",
    "endLine": 5,
    "startLine": 4,
    "variableNames": Array [],
  },
  Object {
    "code": "await page.fill(\\".new-todo\\", todo);",
    "endLine": 7,
    "startLine": 7,
    "variableNames": Array [],
  },
  Object {
    "code": "await page.press(\\".new-todo\\", \\"Enter\\");",
    "endLine": 8,
    "startLine": 8,
    "variableNames": Array [],
  },
  Object {
    "code": "// complete todo
await page.click(
    \\".toggle\\"
);",
    "endLine": 13,
    "startLine": 10,
    "variableNames": Array [],
  },
  Object {
    "code": "// clear completed
await page.click(\\".clear-completed\\");",
    "endLine": 16,
    "startLine": 15,
    "variableNames": Array [],
  },
]
`;

exports[`transformCode takes start and end lines into account 1`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var { page  } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(1)
await page.click(\\"a\\");
};

module.exports = webEditorCode;
"
`;

exports[`transformCode takes start and end lines into account 2`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var { page, x, y, z  } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(5)
await page.click(
  '[placeholder=\\"Search by location..\\"]'
);
};

module.exports = webEditorCode;
"
`;

exports[`transformCode takes start and end lines into account 3`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var { page, x, y, z  } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(5)
await page.click(
  '[placeholder=\\"Search by location..\\"]'
);
vmLineStarted(6)
await page.fill('[placeholder=\\"Search by location..\\"]', \\"Lakeside Village\\");
vmLineStarted(7)
await page.click(\\"a\\");
vmLineStarted(13)
// comment about function
function myFunction() {
  const myInnerConst = \\"hi\\";
  console.log(\\"SUP\\"!);
}
};

module.exports = webEditorCode;
"
`;

exports[`transformCode takes start and end lines into account 4`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var {   } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(1)
var todo = \\"hi there!\\";
variables[\\"todo\\"] = todo;
vmLineStarted(2)
await page.goto('http://todomvc.com/examples/react');
vmLineStarted(5)
// enter new todo
await page.click(\\".new-todo\\");
};

module.exports = webEditorCode;
"
`;

exports[`transformCode takes start and end lines into account 5`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var {   } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(7)
await page.fill(\\".new-todo\\", todo);
vmLineStarted(8)
await page.press(\\".new-todo\\", \\"Enter\\");
};

module.exports = webEditorCode;
"
`;

exports[`transformCode takes start and end lines into account 6`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var {   } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(13)
// complete todo
await page.click(
    \\".toggle\\"
);
vmLineStarted(16)
// clear completed
await page.click(\\".clear-completed\\");
};

module.exports = webEditorCode;
"
`;

exports[`transformCode transforms web editor code 1`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var { page  } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(1)
await page.click(\\"a\\");
};

module.exports = webEditorCode;
"
`;

exports[`transformCode transforms web editor code 2`] = `
"
const webEditorCode = async (variables, { vmEnv, vmLineStarted }) => {
var { page, x, y, z  } = variables;
process.env = { ...process.env, ...vmEnv };
vmLineStarted(5)
await page.click(
  '[placeholder=\\"Search by location..\\"]'
);
vmLineStarted(6)
await page.fill('[placeholder=\\"Search by location..\\"]', \\"Lakeside Village\\");
vmLineStarted(7)
await page.click(\\"a\\");
vmLineStarted(13)
// comment about function
function myFunction() {
  const myInnerConst = \\"hi\\";
  console.log(\\"SUP\\"!);
}
vmLineStarted(15)
myFunction();
vmLineStarted(17)
await myFunction();
vmLineStarted(19)
var myVar = \\"hello\\";
variables[\\"myVar\\"] = myVar;
vmLineStarted(20)
var myConst =  \\"world\\";
variables[\\"myConst\\"] = myConst;
vmLineStarted(21)
var myLet = \\"!\\";
variables[\\"myLet\\"] = myLet;
vmLineStarted(22)
myLet = \\"!!\\";
variables[\\"myLet\\"] = myLet;
};

module.exports = webEditorCode;
"
`;
